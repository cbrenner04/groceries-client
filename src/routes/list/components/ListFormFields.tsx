import React, { type ChangeEventHandler } from 'react';

import { CheckboxField, SelectField, TextField } from 'components/FormFields';
import { EListType } from 'typings';

export interface IListFormFieldsProps {
  name: string;
  type: string;
  completed: boolean;
  refreshed: boolean;
  handleNameChange: ChangeEventHandler;
  handleTypeChange: ChangeEventHandler;
  handleCompletedChange?: ChangeEventHandler;
  handleRefreshedChange?: ChangeEventHandler;
  editForm: boolean;
}

const ListFormFields: React.FC<IListFormFieldsProps> = (props): React.JSX.Element => {
  return (
    <React.Fragment>
      <TextField
        name="name"
        label="Name"
        value={props.name}
        handleChange={props.handleNameChange}
        placeholder="My super cool list"
      />
      <SelectField
        name="type"
        label="Type"
        value={props.type}
        handleChange={props.handleTypeChange}
        options={[
          { value: EListType.BOOK_LIST, label: 'books' },
          { value: EListType.GROCERY_LIST, label: 'groceries' },
          { value: EListType.MUSIC_LIST, label: 'music' },
          { value: EListType.SIMPLE_LIST, label: 'simple' },
          { value: EListType.TO_DO_LIST, label: 'to-do' },
        ]}
        blankOption={false}
      />
      {props.editForm && (
        <React.Fragment>
          {props.handleCompletedChange && (
            <CheckboxField
              name="completed"
              label="Completed"
              value={props.completed}
              handleChange={props.handleCompletedChange}
              classes="mb-3"
            />
          )}
          {props.handleRefreshedChange && (
            <CheckboxField
              name="refreshed"
              label="Refreshed"
              value={props.refreshed}
              handleChange={props.handleRefreshedChange}
              classes="mb-3"
            />
          )}
        </React.Fragment>
      )}
    </React.Fragment>
  );
};

export default ListFormFields;
