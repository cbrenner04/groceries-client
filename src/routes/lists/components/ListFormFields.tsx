import React from 'react';
import PropTypes from 'prop-types';

import { CheckboxField, SelectField, TextField } from '../../../components/FormFields';

function ListFormFields(props) {
  return (
    <>
      <TextField
        name="name"
        label="Name"
        value={props.name}
        handleChange={props.handleNameChange}
        placeholder="My super cool list"
      />
      <SelectField
        name="type"
        label="Type"
        value={props.type}
        handleChange={props.handleTypeChange}
        options={[
          { value: 'BookList', label: 'books' },
          { value: 'GroceryList', label: 'groceries' },
          { value: 'MusicList', label: 'music' },
          { value: 'SimpleList', label: 'simple' },
          { value: 'ToDoList', label: 'to-do' },
        ]}
        blankOption={false}
      />
      {props.editForm && (
        <CheckboxField
          name="completed"
          label="Completed"
          value={props.completed || false}
          handleChange={props.handleCompletedChange || (() => undefined)}
          blankOption={false}
          classes="mb-3"
        />
      )}
    </>
  );
}

ListFormFields.propTypes = {
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  completed: PropTypes.bool,
  handleNameChange: PropTypes.func.isRequired,
  handleTypeChange: PropTypes.func.isRequired,
  handleCompletedChange: PropTypes.func,
  editForm: PropTypes.bool,
};

export default ListFormFields;
