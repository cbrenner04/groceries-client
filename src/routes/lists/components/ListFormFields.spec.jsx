import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import ListFormFields from './ListFormFields';

describe('ListFormFields', () => {
  let props;

  beforeEach(() => {
    props = {
      name: 'foo',
      type: 'GroceryList',
      handleNameChange: jest.fn(),
      handleTypeChange: jest.fn(),
      handleCompletedChange: jest.fn(),
    };
  });

  it('renders new form', () => {
    const { container, getByLabelText, queryByLabelText } = render(<ListFormFields {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByLabelText('Name')).toBeVisible();
    expect(getByLabelText('Type')).toBeVisible();
    expect(queryByLabelText('Completed')).toBeNull();
  });

  it('renders edit form', () => {
    props.editForm = true;
    const { container, getByLabelText } = render(<ListFormFields {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByLabelText('Name')).toBeVisible();
    expect(getByLabelText('Type')).toBeVisible();
    expect(getByLabelText('Completed')).toBeVisible();
  });

  it('calls appropriate change handlers on change of inputs', () => {
    props.editForm = true;
    const { getByLabelText } = render(<ListFormFields {...props} />);

    fireEvent.change(getByLabelText('Name'), { target: { value: 'a' } });

    expect(props.handleNameChange).toHaveBeenCalled();

    fireEvent.change(getByLabelText('Type'), { target: { value: 'a' } });

    expect(props.handleTypeChange).toHaveBeenCalled();

    fireEvent.click(getByLabelText('Completed'));

    expect(props.handleCompletedChange).toHaveBeenCalled();
  });
});
