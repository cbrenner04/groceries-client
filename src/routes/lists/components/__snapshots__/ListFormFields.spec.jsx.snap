// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListFormFields renders edit form 1`] = `
<div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="name"
    >
      Name
    </label>
    <input
      class="form-control"
      id="name"
      name="name"
      placeholder="My super cool list"
      type="text"
      value="foo"
    />
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="type"
    >
      Type
    </label>
    <select
      class="form-control"
      id="type"
      name="type"
    >
      <option
        value="BookList"
      >
        books
      </option>
      <option
        value="GroceryList"
      >
        groceries
      </option>
      <option
        value="MusicList"
      >
        music
      </option>
      <option
        value="SimpleList"
      >
        simple
      </option>
      <option
        value="ToDoList"
      >
        to-do
      </option>
    </select>
  </div>
  <div
    class="mb-3 mb-3"
  >
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="completed"
        name="completed"
        type="checkbox"
      />
      <label
        class="form-check-label"
        for="completed"
        title=""
      >
        Completed
      </label>
    </div>
  </div>
</div>
`;

exports[`ListFormFields renders new form 1`] = `
<div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="name"
    >
      Name
    </label>
    <input
      class="form-control"
      id="name"
      name="name"
      placeholder="My super cool list"
      type="text"
      value="foo"
    />
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="type"
    >
      Type
    </label>
    <select
      class="form-control"
      id="type"
      name="type"
    >
      <option
        value="BookList"
      >
        books
      </option>
      <option
        value="GroceryList"
      >
        groceries
      </option>
      <option
        value="MusicList"
      >
        music
      </option>
      <option
        value="SimpleList"
      >
        simple
      </option>
      <option
        value="ToDoList"
      >
        to-do
      </option>
    </select>
  </div>
</div>
`;
